lib/rubocop/cop/rails/duplicate_ignored_columns.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cops looks for assignments of `ignored_columns` that override previous
+      # assignments.
+      #
+      # Overwriting previous assignments is usually a mistake, since it will
+      # un-ignore the first set of columns
+      #
+      # @example
+      #
+      #   # bad
+      #   class User < ActiveRecord::Base
+      #     self.ignored_columns = [:one]
+      #     self.ignored_columns = [:two]
+      #   end
+      #
+      #   # bad
+      #   class User < ActiveRecord::Base
+      #     self.ignored_columns += [:one]
+      #     self.ignored_columns = [:two]
+      #   end
+      #
+      #   # good
+      #   class User < ActiveRecord::Base
+      #     self.ignored_columns = [:one, :two]
+      #   end
+      #
+      #   # good
+      #   class User < ActiveRecord::Base
+      #     self.ignored_columns = [:one]
+      #     self.ignored_columns += [:two]
+      #   end
+      #
+      class DuplicateIgnoredColumns < Base
+        MSG = 'This assignment to `ignored_columns` overwrites previous ones.'
+
+        def_node_matcher :ignored_columns_assign, <<-PATTERN
+          (send self :ignored_columns= ...)
+        PATTERN
+        def_node_matcher :ignored_columns_append, <<-PATTERN
+          (op-asgn
+            (send self :ignored_columns)
+            :+ ...
+          )
+        PATTERN
+
+        def on_op_asgn(node)
+          ignored_columns_append(node) do
+            @seen_before = true
+          end
+        end
+
+        def on_send(node)
+          ignored_columns_assign(node) do
+            add_offense(node) if @seen_before
+            @seen_before = true
+          end
+        end
+
+        def on_new_investigation
+          @seen_before = false
+          super
+        end
```

Comment:
Maybe `on_class` (and `on_module`, too?) and check `node.each_children.select(&method(:ignored_columns_assign?))` to replace all three methods?

Or, if we decide to bad all `ignored_columns =` and only allow for `+=`, `on_send` would be sufficient.

Reply:
Hmm. The official docs mention using `=`, not `+=`. https://api.rubyonrails.org/classes/ActiveRecord/ModelSchema/ClassMethods.html#method-i-ignored_columns

It would be annoying if rubocop complained by default on standard usage

Reply:
If Rails doesn't warn on duplicate in runtime and `+=` is safe by default, I don't see a reason why it shouldn't be a good practice, even though it's not exactly standard. And I guess it's not standard not to overcomplicate things for a rare misuse.

Reply:
Interestingly, today a conversation surfaced at work about this, and it involved setting `ignored_columns` in STI classes, which override the base class. Thus, I'll update to follow your suggestion to only allow `+=`.

Reply:
Fantastic!

Just one question - why unsafe for auto-correction exactly? It's a late hour here and I might miss something.
Do you mean that models will change its behaviour due to the past presence of an erroneous override, and its later disappearance? How impactful can be that it would suddenly start to ignore certain columns?

Reply:
Because avoiding an overwrite is possibly breaking for the app (although *very* unlikely). My interpretation of `safe` was that the rewritten code has no discernible difference to the original code. `a += b` can be rewritten as `a = a + b` because both expressions are strictly the same. But `a.concat(b)` cannot be (safely) rewritten to `a += b` because the side effect of modifying `a` is now gone. But perhaps my intepretation is too restrictive?

Reply:
I think that the effect of properly ignoring columns will be detected by running specs.
However, formally you're right (https://docs.rubocop.org/rubocop/1.24/usage/auto_correct.html#safe-auto-correct):
> SafeAutoCorrect (true/false) - indicates whether the auto-correct a cop does is safe (**equivalent**) by design.

Reply:
So, do I leave it as unsafe or change to safe?

Reply:
`SafeAutoCorrect: false` is fine 👍